### PR TITLE
Add PropertyType tests

### PR DESCRIPTION
**Pull request by zach-albia**
#169

The core idea is to test that string values read, then written by a `PropertyType[String, A]` instance and finally parsed to an `A`, yield the same result as parsing to an `A`. That is, reading from and writing to string values doesn't change the value of the properties they parse to.

Each `PropertyType` instance, except `StringType`, has a valid and invalid string property test.

Generators summary:
- String generators for all numeric types are padded with (0..*) leading zeros for added coverage.
- All case permutations of the strings `"true"` and `"false"` are used to test `BooleanType`.
- `BigInt`s and `BigDecimal`s are generated using numeric strings.....is this okay?
- I based the `BigDecimal` generator loosely on [this diagram](https://i.stack.imgur.com/wmFqa.gif).
- I just reused `genDuration` from helpers.scala.
- Generated URI strings are generic, i.e. loosely based on the [spec here](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Generic_syntax).
- I make heavy use of a `genAppend`, which stitches `Gen[R, String]` values together, which I think should be written in terms of a `Gen.sequence`, as per ZIO issue [#2573](https://github.com/zio/zio/issues/2573).

### CODE REVIEW

**Comment by zach-albia** [comment]
Whoops, forgot to test on 2.12.x. Also, the genUri is buggy. Fixing...

File: core/src/test/scala/zio/config/PropertyTypeTest.scala [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+          propType = DoubleType,
+          genValid = genPadLeadingZeros(Gen.double(Double.MinValue, Double.MaxValue)),
+          parse = _.toDouble
+        ),
+        propertyTypeRoundtripSuite(
+          typeInfo = "BigDecimal",
+          propType = BigDecimalType,
+          genValid = genValidBigDecimalString,
+          parse = BigDecimal(_)
+        ) @@ flaky, // because nums generated could be too large
```
Comment (jdegoes) [inline core/src/test/scala/zio/config/PropertyTypeTest.scala]: You could solve this by generating only those in a certain range.

File: core/src/test/scala/zio/config/PropertyTypeTest.scala [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+  private val genDigit0To9: Gen[Random, String] = char('0', '9').map(_.toString)
+  private val genDigit1To9: Gen[Random, String] = char('1', '9').map(_.toString)
+
+  /**
+   * Generates a string of up to n digits. Default maxLength digits is 100.
+   */
+  private def genDigits(maxLength: Int = 100): Gen[Random with Sized, String] =
+    sized(n => listOfN(math.min(n, maxLength))(genDigit0To9).map(_.mkString))
+
```
Comment (zach-albia) [inline core/src/test/scala/zio/config/PropertyTypeTest.scala]: @jdegoes Limiting digits to 100 by default. I went and checked the max and min numbers for this and both `BigInt` and `BigDecimal` and both are fine with them.

**Comment by afsalthaj** [comment]
@jdegoes Are we good to merge this one?

**Pull request merged**